homeassistant/helpers/event.py

```diff
@@ -90,34 +99,246 @@ def state_change_listener(event):
+    template
+        The template to calculate.
+    action
+        Callable to call with results. See above for arguments.
+    variables
+        Variables to pass to the template.
+
+    Returns
+    -------
+    Callable to unregister the listener.
+
+    """
+    @callback
+    def state_changed_listener(event, template, last_result, result):
+        """Check if condition is correct and run action."""
+        if isinstance(result, TemplateError):
+            _LOGGER.exception(result)
+            return
+
+        if last_result is not None:
+            if not _boolean_coerce(last_result) \
+                    and _boolean_coerce(result):
+                hass.async_run_job(action, event.data.get('entity_id'),
+                                   event.data.get('old_state'),
+                                   event.data.get('new_state'))
+        elif _boolean_coerce(result):
+            # First run of the listener. Figure out an entity ID to
+            # pass back to the action because it expects one.
+            info = template.async_render_to_info(variables)
+            state = None
+            entity_id = None
+            #  pylint: disable=protected-access
+            for eid in info._entities:
+                state = hass.states.get(eid)
+                if state:
+                    entity_id = eid
+                    break
+            if not state:
+                for st in hass.states.async_all():
+                    if info.filter_lifecycle(st.entity_id):
+                        state = st
+                        entity_id = st.entity_id
+                        break
+            hass.async_run_job(
+                action, entity_id or MATCH_ALL, state, state)
+
+    info = async_track_template_result(
+        hass, template, state_changed_listener, variables)
+
+    return info.async_remove
+
+
+track_template = threaded_listener_factory(async_track_template)
+
+
+_UNCHANGED = object()
+
+
+class TrackTemplateResultInfo:
+    """Return value from async_track_template_result."""
+
+    def __init__(self, hass, template, action, variables):
+        """Initialiser, should be package private."""
+        self.hass = hass
+        self._template = template
+        self._action = action
+        self._variables = variables
 
-    # Local variable to keep track of if the action has already been triggered
-    already_triggered = False
+        self._cancel = hass.bus.async_listen(
+            EVENT_STATE_CHANGED, self._state_changed_listener)
+
+        self._info = template.async_render_to_info(variables)
+        try:
+            self._last_result = self._info.result
+        except TemplateError as ex:
+            self._last_result = None
+            self._last_exception = True
+            self.hass.async_run_job(
+                self._action, None, self._template,
+                None, ex)
+        else:
+            self._last_exception = False
+            self.hass.async_run_job(
```

Comment:
Why are we running the action as part of the initialization?

Reply:
If this is intentional, this should be `async_add_job` instead of `async_run_job`, because if people call this as part of some initialization, their callback might not be fully initialized yet. With async_add_job, they can finish initialization and the callback won't be called until the next cycle of the event loop.

Reply:
It's intentional yes, otherwise you end up evaluating the template outside, then running the action anyhow, then sending the template to the listener that then runs the template again so it can determine what entities that it touches, and what the result is to see if it's changed. I couldn't find anywhere where it was actually not done like that so it made more sense just to immediately fire if required.

Agree to the `async_add_job` though.

Reply:
I'm not sure if I follow. All methods in this file are helpers to track when something happens that is interesting. Registering a listener should not call something, we should leave that to the registering code to decide if that is something they want to do. By making that decision, there is no opt-out. By removing that it is called directly, we also don't have to include a fake entity ID.

Reply:
I agree that's the better way to have it in terms of what would normally be considered this method's jobs.

The issues were: 
* every place that does call `async_track_template_result` was effectively just calling the registration method then immediately calling the listener anyhow so it seemed simpler just to immediately call the listener on registration.
* Calling the template to evaluate it first, then calling the registration method would require an extra template evaluation, so that didn't make a lot of sense.
* `async_track_template` did still immediately call the listener so I kept this one doing the same thing.

Another alternative would be to have a `last_result_info` property on the returned `TrackTemplateResultInfo` object, and let the caller get the data if it wants to. That makes more sense to me in terms of responsibilities, although it does come at a cost of boiler-plate code in places that call this method.

Happy to change it to this last paragraph if you think it's better (and I actually agree, just there's a trade-off that's all).

Reply:
Will there ever be a use case where we do not want it to be called right away?

Reply:
There isn't any in the code base currently, and I can't think of a good reason why that would happen. If there was one that came up then you could just discard the first call rather than having to change every other place that uses it to suit that case. You still need to evaluate the template regardless, it's more a matter of if you then pass this result back to the caller.

There's cases when no particular entity causes the fire, like in manual refreshes of the template entities, so they need to be handled too, so it's not like you're introducing potential bugs. If anything the first callback should have no associated event or entity ID.

Reply:
Maybe I am not 100% following, but for example for automations, we wouldn't want our trigger to be called right away. Nor does this behavior exist in any other event listener in this file.

Reply:
It has been a while since I've looked admittedly, for templates you need it to fire immediately so that you can use the current value. Shouldn't automations fire at start-up time? 

Anyhow happy to go with the `last_result_info` option from above if you think that's better.

Reply:
No automation fires at startup time. That would be dangerous if all your automations fire at once 😉 They should only fire if a certain condition has been met.

Reply:
We can go with `last_result_info` then.

Reply:
I still have not seen a convincing case to call the listener when the listener is attached. I don't think that we should merge that.

Reply:
Yep, will change it. I won't get any time to to anything until next week though.

Reply:
No rush, no new beta for another 2 weeks !

Reply:
Bump

Reply:
Yep I know, time poor at the moment :slightly_frowning_face:. Will work on it next Wednesday.